### DISCUSSION TITLE
# Day 7 · Empty vectors and FAISS pitfalls (No.8)

### DISCUSSION THREAD

**Discussion by onestardao** [onestardao/WFGY#32] [General]
# Day 7 · Empty vectors and FAISS pitfalls (No.8)

> Symptom  
> ingestion says “ok”, indexes look built, but recall is dead. queries return junk or the same ids over and over. everything feels fine until you actually measure neighbors.

**ProblemMap**

* **No.8 Debugging is a Black Box**
  reference map lives in the project root

---

## What this usually looks like

* FAISS `index.ntotal` equals your doc count, yet `search` returns near identical ids for unrelated queries
* recall@k on a small smoke set is near zero
* IVF shows many busy lists but re-rank separation is poor
* you swap models or re-embed, then recall falls off a cliff although ingestion printed “done”
* HNSW searches are shallow and top-k barely changes with the query

---

## 60-sec quick test

1) take a 5k sample of your embeddings  
2) compute row norms, zero counts, NaN or Inf counts  
3) check metric and normalization policy match the index you built

Rules of thumb

* **zero vectors must be 0.0%**. anything above 0 is a hard red flag  
* **NaN or Inf must be 0**. if not, drop the whole batch and re-embed  
* for cosine, vectors should be L2-normalized. for L2, normalization is optional but norms should not all be exactly 1.0  
* if top-k for ten random queries shares more than **35%** of ids, something is off in geometry or ingestion

---

## Diagnose checklist

* tokenizer or parser produced empty text spans, then your embedder returned zeros
* dtype or device mistake during batch embed wrote all zeros to disk
* shape mismatch while saving restored only a prefix of the vector (wrong `d`)
* metric mismatch. IP index with already normalized vectors or L2 over non-normalized vectors
* IVF or PQ codebooks reused after whitening or model swap
* mixed shards with different normalization or different metrics
* query side got normalized but corpus side did not, or the other way around
* id collisions. you silently overwrote rows and kept `ntotal` correct but lost evidence
* ingestion order wrong. you trained IVF before dedup or before masking boilerplate

---

## Minimal fix

**Goal** make the retriever trustworthy again with the smallest change set.

1. **Reject bad rows before index add**

```python
   import numpy as np

   assert embs.ndim == 2, "bad shape"
   norms = np.linalg.norm(embs, axis=1)
   bad_zero = np.where(norms == 0)[0]
   bad_nan  = np.where(~np.isfinite(norms))[0]
   if len(bad_zero) or len(bad_nan):
       raise RuntimeError(f"zero:{len(bad_zero)} naninf:{len(bad_nan)}")
````

2. **Align metric with vector state**

   * cosine retrieval → L2-normalize both corpus and queries, then use L2 or IP consistently
   * inner product → do not re-normalize twice, control norms explicitly

3. **Rebuild the index from clean embeddings**
   do not patch mixed shards. trash and rebuild

4. **Run a 5-question smoke test**
   fixed questions with known spans. if recall stays low, stop and re-check geometry

---

## Hard fixes when minimal is not enough

* **Re-train IVF or PQ codebooks** on a large and diverse clean sample
* **Corpus hygiene**. dedup near duplicates and mask boilerplate before embedding
* **Dimension and contract check**. assert `embs.shape[1] == expected_d` at every hop
* **One policy per store**. one metric, one normalization policy, one whitening state
* **Do not mix OPQ and non-OPQ** in a single store

---

## WFGY guardrails you can enable

* **Trace schema** for No.8. log decisions, not prose, at each hop
* **Bridge step** when retrieval returns thin evidence. restate last valid state and required constraint before you continue generation
* **Failure counters**. zero-vector rate, NaN rate, neighbor-overlap, recall. block deploy when a counter crosses threshold

> in forums, label this class as **No.8**. if boot order contributed, also note **No.14**. if you tested against an empty or mixed store, tag **No.16**.

---

## Tiny scripts you can paste

**zero / NaN / dim sanity**

```python
import numpy as np

def sanity(embs, expected_d):
    assert embs.ndim == 2 and embs.shape[1] == expected_d, f"dim mismatch {embs.shape}"
    norms = np.linalg.norm(embs, axis=1)
    z = int(np.sum(norms == 0))
    nf = int(np.sum(~np.isfinite(norms)))
    return {"rows": embs.shape[0], "zero": z, "naninf": nf,
            "min_norm": float(norms.min()), "max_norm": float(norms.max())}

# usage
# stats = sanity(embs, 768); print(stats)
```

**metric match + FAISS rebuild (cosine via L2)**

```python
import faiss, numpy as np
from sklearn.preprocessing import normalize

Z = np.load("embeddings.npy").astype("float32")
Z = normalize(Z, axis=1)           # cosine requires L2-normalized
faiss.normalize_L2(Z)              # safety

d = Z.shape[1]
index = faiss.IndexHNSWFlat(d, 32) # HNSW with L2 acts like cosine when vectors are normalized
index.hnsw.efConstruction = 200
index.add(Z)
faiss.write_index(index, "hnsw_cosine.faiss")
```

**neighbor-overlap sanity**

```python
import numpy as np

def overlap_at_k(nei_a, nei_b, k=20):
    a, b = set(nei_a[:k]), set(nei_b[:k])
    return len(a & b) / float(k)

# run on 10 random queries, average overlap. healthy spaces sit well below 0.35
```

**IVF/PQ retrain sketch**

```python
import faiss, numpy as np

Z = np.load("embeddings_clean.npy").astype("float32")
nlist, m, nbits = 4096, 64, 8

quant = faiss.IndexFlatL2(Z.shape[1])
ivfpq = faiss.IndexIVFPQ(quant, Z.shape[1], nlist, m, nbits)
ivfpq.train(Z)         # train on a large, clean, deduped sample
ivfpq.add(Z)
ivfpq.nprobe = 32
faiss.write_index(ivfpq, "ivfpq_l2.faiss")
```

---

## Acceptance checks before you call it fixed

* zero-vector rate and NaN rate are both **0.0%**
* metric and normalization policy are documented and match the index type
* PC1 explained variance sits in a healthy band after whitening if you use cosine
* neighbor-overlap across 20 random queries is **≤ 0.35** at k 20
* recall\@k on a heldout set rises, and top-k lists actually change with the query
* `index.ntotal` equals the number of valid rows you ingested. no silent drops

---

## TL;DR

ingestion “ok” is not proof. kill zero vectors and NaNs, align metric to vector state, rebuild from clean data, then smoke test. if retrieval is thin, add a bridge step so the chain says what is missing before it continues. call it **No.8** and keep traces.

---

[https://github.com/onestardao/WFGY/blob/main/ProblemMap/article/README.md](https://github.com/onestardao/WFGY/blob/main/ProblemMap/article/README.md)
